shell/server/hack-session.js

```diff
@@ -323,10 +323,26 @@ HackSessionContextImpl.prototype.generateApiToken = function (petname, userInfo,
 };
 
 Meteor.methods({
-  newApiToken: function (grainId, petname, roleAssignment, forSharing, destroyIfNotUsedByTime) {
-    // Create a new user-oriented API token.
-    if (!this.userId) {
-      throw new Meteor.Error(403, "Must be logged in to create an API token.");
+  newApiToken: function (grainId, petname, roleAssignment, forSharing, expiresIfUnusedDuration,
+                         rawParentToken) {
+    // Creates a new UiView API token. If `rawParentToken` is set, creates a child token.
+    var userId = this.userId;
+    var parentToken;
+    if (!userId) {
+      if (!rawParentToken) {
+        throw new Meteor.Error(403, "If you are not logged in, you must specify a parent token to " +
+                                    "create a new API token.");
+      } else {
+        check(rawParentToken, String);
+        parentToken = Crypto.createHash("sha256").update(rawParentToken).digest("base64");
+        var parentApiToken = ApiTokens.findOne({_id: parentToken, grainId: grainId,
+                                                objectId: {$exists: false}});
+        if (!parentApiToken) {
+          throw new Meteor.Error(403, "No such parent token found.");
+        }
+        userId = parentApiToken.userId;
+        forSharing = parentApiToken.forSharing;
```

Comment:
If the parent has (forSharing = false) and a user creates a new token using the hypothetical "Asheesh button" (i.e. the "remember token" button) in the webkey menu, then it would make sense for the child to have (forSharing = false). However, maybe that's not an important case now that we have offer templates.